### PR TITLE
host_build_graph: retire last_task_alive and the dead reclamation path

### LINKED ISSUE #1720: [Code Health] hbg: Retire last_task_alive and the dead mid-run reclamation path
### Category

Dead Code / Unused Dependencies

### Component

Ring Buffer

### Description

`host_build_graph` still carries the complete mid-run reclamation subsystem it
inherited from `tensormap_and_ringbuffer`, driven by a shared-memory field that
**nothing in HBG ever writes**.

`PTO2RingFlowControl::last_task_alive` is defined as the orchestrator↔scheduler
back-pressure channel — its own comment says *"Cache Line 1: Written by Scheduler,
Read by Orchestrator (for back-pressure)"*. In T&R the on-device orchestrator and the
schedulers share the ring, and the scheduler advancing this field is what releases
task slots and heap bytes for reuse.

HBG is whole-graph-resident and reclaims nothing during a run, so the scheduler
deliberately does not advance it — `pto_scheduler.h:26`: *"there is no on-device slot
reclaim (whole-graph-resident), so `last_task_alive` is not advanced here."* Outside
`init()` there is no writer in the tree. It is pinned at 0 for the whole run.

The host orchestrator nevertheless reads it on every submit and drives an entire
chain of code that can only ever be a no-op:

| Site | What it does with a permanently-zero value |
| ---- | ------------------------------------------ |
| `pto_orchestrator.cpp:1169`, `:1494` | `sync_tensormap(task_id, 0)` → `sync_validity(0)` and a periodic `cleanup_retired(0, 0)` that can never free an entry |
| `pto_tensormap.h:668` | `is_entry_valid` compares `producer_local >= last_task_alive_cached`, i.e. `>= 0` — unconditionally true |
| `pto_orchestrator.cpp:1191-1192` | `dep_local_task_id < dep_last_task_alive`, the "producer already retired, no fanin wiring needed" shortcut — unreachable |
| `pto_ring_buffer.h:159` | `// Spin: wait for scheduler to advance last_task_alive` — a spin whose exit condition can never be met |
| `pto_ring_buffer.h:272` | `last_alive_seen_` / `update_heap_tail()`, the heap-side half of the same reclaim |
| `pto_runtime2_init.cpp:268` | passes `last_alive_dev` into `PTO2TaskAllocator::init` |

Two things make this worth removing rather than tolerating.

**It costs work on the hot build path.** Every submit pays `sync_tensormap`, an epoch
compare, and a periodic `cleanup_retired` walk that provably frees nothing. Host build
time is on HBG's critical path, so this is not free.

**It produces a misleading diagnosis on the failure that matters most.** When the graph
does not fit, the allocator enters the spin above, waits for a reclaim that can never
arrive, and the §4.1 wall-clock backstop eventually latches
`FATAL: Task Allocator Deadlock - Heap Exhausted!`. There is no deadlock: the graph
simply exceeds the configured task window or heap. Anyone triaging that message starts
by looking for a dependency cycle.

HBG already has the signal that actually matters for consumer retirement, and it is a
different field: `completed_watermark`, which the scheduler advances and the host's
`wait_for_consumers` gates on (`watermark >= producer.last_consumer_local_id`). Nothing
about removing `last_task_alive` touches it.

Note what must **not** be removed: the TensorMap lookup itself is how HBG derives
dependencies, and the heap allocator's *allocation* cursor is how packed buffers get
addresses. Only the retirement/reclaim halves are dead.

### Location

```markdown
- `src/a2a3/runtime/host_build_graph/runtime/pto_shared_memory.h:55` — the field, plus `static_assert(sizeof(PTO2RingFlowControl) == 128)` that its removal changes
- `src/a2a3/runtime/host_build_graph/runtime/pto_shared_memory.h:297` — `ring_last_task_alive_addr`
- `src/a2a3/runtime/host_build_graph/runtime/pto_ring_buffer.h:159`, `:272` — the never-satisfiable spin and `last_alive_seen_` / `update_heap_tail`
- `src/a2a3/runtime/host_build_graph/runtime/pto_tensormap.h:378`, `:394-405`, `:492`, `:582-590`, `:668` — `last_task_alive_cached`, `reclaim_retired_all`, `sync_validity`, `cleanup_retired`, `is_entry_valid`
- `src/a2a3/runtime/host_build_graph/runtime/shared/pto_tensormap.cpp:211-220` — `sync_tensormap` and `PTO2_TENSORMAP_CLEANUP_INTERVAL`
- `src/a2a3/runtime/host_build_graph/runtime/orchestrator_core/pto_orchestrator.cpp:1041`, `:1169`, `:1191-1192`, `:1494` — the reads and the retired-producer shortcut
- `src/a2a3/runtime/host_build_graph/runtime/shared/pto_runtime2_init.cpp:92`, `:268` — init and the allocator constructor argument
- `src/a5/runtime/host_build_graph/` — the same sites exist in the a5 tree
```

### Proposed Fix

Remove the field and every consumer listed above, in both arch trees in one commit.

**This changes one behavior and the change is the point:** a graph that does not fit
the configured task window, heap, fanin capacity or TensorMap pool must fail
immediately with a capacity error naming the exhausted resource and the requested vs.
available amount, instead of spinning until the wall-clock backstop reports a
"deadlock". Update the §4.1 text in `RUNTIME_LOGIC.md` accordingly — the representative
allocator output quoted there changes.

`PTO2_ERROR_TENSORMAP_OVERFLOW`'s description (*"tensormap entry pool wedged
(last_task_alive not advancing)"*) also needs rewording, since after this change there
is no such thing as a pool that could have been unwedged by reclaim.

Sequencing: this touches `pto_runtime2_init.cpp` and `scheduler_dispatch.cpp` regions
that the multi-ring collapse also touches. Land this before that one, or rebase.

### Priority

Medium (minor risk, should fix in next few releases)

### PR DESCRIPTION
**Pull request by ChaoZheng109**
Fixes #1720

## What

`host_build_graph` inherited `tensormap_and_ringbuffer`'s mid-run reclamation
subsystem, driven by `PTO2RingFlowControl::last_task_alive`. In T&R the on-device
scheduler advancing that field is what frees task slots and heap bytes. **hbg is
whole-graph-resident and reclaims nothing during a run**, so the scheduler
deliberately never advances it (`pto_scheduler.h:26`) and it is pinned at 0 —
while the host orchestrator read it on every submit and drove a chain of provable
no-ops:

| site | with a permanently-zero watermark |
| ---- | --------------------------------- |
| `sync_tensormap()` per submit | `sync_validity(0)` + a `cleanup_retired(0, 0)` gate that can free nothing |
| `entry_valid()` per TensorMap chain step | `producer_local >= 0` — unconditionally true |
| retired-producer fanin shortcut | `dep_local < dep_last_alive` unreachable, so every declared dep pays full fanin wiring |
| allocator `alloc()` | spun on a reclaim that cannot arrive, until a 500 ms backstop latched `Task Allocator Deadlock - Heap Exhausted` |

This removes the field and the reclaim halves it drove. Both rings become
forward-only bump allocators: `heap_tail_`, `last_alive_seen_`, the heap rebase
anchor, `update_heap_tail()` and `try_bump_heap()`'s wrap branches are all
unreachable without reclaim, as are the `scope_stats` heap-wrap reports and the
`descriptors_` / `slot_states_` pointers only the reclaim path read.

## Behavior change — this is the point

A graph that does not fit the configured task window, heap or TensorMap pool can
never become satisfiable by waiting. Allocation now fails on the spot and names
the exhausted resource, instead of reporting a deadlock 500 ms later. Verified
onboard with `PTO2_RING_TASK_WINDOW=64`:

```text
FATAL: Task Window Exhausted!
The whole graph must fit the configured ring; nothing is reclaimed mid-run.
  Task window: used=63/64
  Graph heap:  used=10481664/268435456, available=257953792
  Requested:   8192 bytes + 1 task slot
Solution:
  Increase task window (current: 64); env PTO2_RING_TASK_WINDOW=<pow2> (e.g. 128)
```

No `No reclaim progress` / `Provable head-of-line` / `Allocator Deadlock` wording
remains on this path.

## Two scoping calls worth a reviewer's eye

**Numeric error codes are unchanged.** `PTO2_ERROR_HEAP_RING_DEADLOCK` /
`FLOW_CONTROL_DEADLOCK` now read oddly for hbg, but the names live in
`src/common/runtime_status/error_names.h`, shared with T&R where the deadlock
reading is still accurate. Renaming them needs a coordinated cross-runtime change.
`orch_mark_fatal` is first-writer-wins, so the resource-specific code the
allocator latches survives `prepare_task`'s generic follow-up — the surfaced code
is correct, only its *name* is now imprecise for hbg.

**`initial_local_task_id` is gone from `PTO2TaskAllocator::init`.** The old window
check was `local_task_id_ - last_alive + 1 < window_size_`. Without a watermark
the choices were to keep a local `base_task_id_` (`last_task_alive` under another
name) or to fix the ring's origin at 0. Fixed at 0: with no reclaim,
`local_task_id_ < window_size_` is a hard invariant, so ids provably cannot
approach `INT32_MAX` and the seeded corner case has no meaning for this runtime.

## Tests

`tests/ut/cpp/a2a3/test_task_allocator.cpp` is compiled **twice** — against T&R as
`test_task_allocator` and against hbg as `test_hbg_task_allocator` — and drives
`last_alive` in 27 places including the `init()` call, so it cannot serve both
after this change. The hbg target moves to a new
`tests/ut/cpp/a2a3/test_hbg_task_allocator.cpp` covering the forward-only contract
and immediate capacity failure. `test_hbg_tensormap.cpp` swaps its
`cleanup_retired` cases for entry-lifetime and pool-occupancy ones. **The T&R
source and target are untouched**, and `test_task_allocator` still passes.

## Verification

| | result |
| --- | --- |
| a2a3 + a5 hbg `aicpu`/`host` builds | clean; the shifted SM-layout `static_assert`s check the new offsets |
| `cpput` | 97/97 pass, including T&R `test_task_allocator` |
| `a2a3sim` / `a5sim` scene tests | exit 0, zero FAILED/ERROR |
| a2a3 onboard sweep (CI args, 4 dies) | exit 0, 31 + 57 passed, 1 skipped, zero FAILED |
| a2a3 onboard `host_build_graph/qwen3_14b_decode` (level 4, golden on) | 1 passed, 125 s |
| capacity failure path | immediate, named resource (above) |
| pre-commit | all hooks pass |

**Perf**: qwen3-14B decode, die 6, 100 rounds each, warmup dropped, 1.5×IQR
filtered — before vs after. No regression, and no claimable win: this workload
records one decoder layer as a Graph and replays it 39 times, so it submits only
47 tasks per round and a per-submit saving cannot clear the noise floor.

| span (ms) | before | after | Δ |
| --- | --- | --- | --- |
| `simpler_run` | 2042.9 | 2027.4 | −15.5 |
| `simpler_run.bind` | 1281.3 | 1273.4 | −7.9 |
| `runner_run` | 45.63 | 45.55 | −0.09 |
| `device_wall` | 44.17 | 44.18 | +0.01 |

Run-to-run sd on `bind` is 24–34 ms, so both deltas are noise. The value here is
code health plus the corrected failure diagnosis, not throughput.

## Sequencing

Land **before** #1721 (multi-ring collapse) — it touches `pto_runtime2_init.cpp`,
`pto_shared_memory.cpp` and `runtime_maker.cpp` broadly enough to conflict.
#1719 is independent of both. Both arch trees move together per #1706.

### CODE REVIEW

**Comment by zmnobug** [comment]
I reviewed this against the full scope of #1720 and the merge-base diff. The overall direction is sound, but I do not think it is ready to merge yet.

### Blocking issues

1. **The final task-window slot is rejected.**

   `PTO2TaskAllocator::alloc()` checks `local_task_id_ + 1 >= window_size_`, so a window of 64 only admits task IDs 0..62. The new `TaskWindowSaturates` test encodes the same `WINDOW_SIZE - 1` expectation, which is why CI stays green.

   This needs to be fixed consistently in both arch trees and all admission paths:
   - `pto_ring_buffer.h`: allow allocation while `local_task_id_ < window_size_`
   - `graph_submit_definition`: fix the matching preflight
   - `check_scope_can_accept_task`: it still rejects at `window_size - 1` and reports `Scope Deadlock`, so changing only the allocator will not make the last slot usable

   The regression test should prove that the first N allocations succeed, allocation N+1 fails immediately, and `current_index` remains unchanged on failure.

2. **The fanin-capacity part of #1720 is missing.**

   The issue explicitly requires immediate, resource-specific capacity diagnostics for task window, heap, fanin, and TensorMap. The PR body and implementation cover only task window, heap, and TensorMap.

   `append_fanin_or_fail()` still only latches `PTO2_ERROR_DEP_POOL_OVERFLOW` at `PTO2_MAX_FANIN`, with no fanin limit/requested diagnostic. The shared hint then recommends `PTO2_RING_DEP_POOL`, even though HBG has no dependency spill pool. Please add the HBG fanin-capacity diagnostic and a regression test.

3. **The surfaced error descriptions still contradict the new behavior.**

   `src/common/runtime_status/error_names.h` still says:
   - code 2 means task slots and heap are both exhausted, while the new allocator uses it for heap-only exhaustion;
   - code 3 is a nested flow-control deadlock, while the new allocator uses it for task-window capacity;
   - code 4 is a dependency spill-pool overflow, which is not the HBG fanin model;
   - code 11 is a TensorMap pool wedged by `last_task_alive`, even though this PR removes that field.

   The code 11 description update is an explicit #1720 requirement, and its current hint says capacity tuning will not help while the new runtime message says to increase the TensorMap pool. Numeric codes can remain shared, but the description/hint path must not give HBG users contradictory advice.

### Follow-ups

Please also clean up the stale reclamation comments in both `pto_tensormap.h` copies, the deadlock-detector comment above `ring_slot_states_addr`, and the now-unused HBG `PTO2_DEP_POOL_CLEANUP_INTERVAL`. The TensorMap slot-alias test should assert both producer IDs, not only the result count.

All CI checks currently pass, but the window-boundary UT currently asserts the buggy N-1 capacity, so green CI does not cover the blocking issue above. The branch is also 25 commits behind current `main`; after fixing/rebasing, the full matrix should be rerun.

File: .claude/rules/running-onboard.md [hunk 1]
```diff
@@ -174,7 +174,8 @@ for the signature that actually fired:**
...
-| `No reclaim progress for ~500 ms` / `cannot reclaim space after ~500 ms` | allocator/pool **reclaim timeout** | The 500ms backstop (`PTO2_ALLOC_DEADLOCK_TIMEOUT_CYCLES`) exists in HBG and TRB. It proves prolonged lack of reclaim progress, not why progress stopped; check capacity, the dumped head, consumers and scheduler state. |
+| `No reclaim progress for ~500 ms` / `cannot reclaim space after ~500 ms` | allocator/pool **reclaim timeout** | TRB only — the 500ms backstop (`PTO2_ALLOC_DEADLOCK_TIMEOUT_CYCLES`) lives in its reclaiming allocators. It proves prolonged lack of reclaim progress, not why progress stopped; check capacity, the dumped head, consumers and scheduler state. |
+| `Task Window Exhausted` / `Graph Heap Exhausted` / `TensorMap Entry Pool Exhausted` | **capacity**, HBG only | HBG is whole-graph-resident and reclaims nothing mid-run, so this is a sizing verdict reached immediately, not a stall. The line carries used/capacity and the requested amount; raise `PTO2_RING_TASK_WINDOW` / `PTO2_RING_HEAP` / `PTO2_TENSORMAP_POOL_SIZE` or shrink the graph. |
```
[the remark this answers is not in the document]
Comment (zmnobug) [reply .claude/rules/running-onboard.md]: Fixed in e7e6b96a. HBG now emits `FATAL: Fanin Capacity Exhausted!` with used/capacity and the requested distinct-producer count. The diagnostic and troubleshooting docs identify `PTO2_MAX_FANIN=128` as the hard HBG limit and explicitly state that `PTO2_RING_DEP_POOL` does not apply. The shared code-4 hint now distinguishes HBG inline fanin from the reclaiming T&R dependency pool.

File: src/a2a3/runtime/host_build_graph/runtime/pto_ring_buffer.h [hunk 2]
```diff
@@ -121,184 +101,87 @@ class PTO2TaskAllocator {
...
-                        blocked_on_heap ? "heap" : "task", spin_count
-                    );
-                }
-            }
-            SPIN_WAIT_HINT();
+        // Check both resources; commit only if both are available.
+        if (local_task_id_ + 1 >= window_size_) {
+            report_capacity_exhausted(/*heap_blocked=*/false, aligned_size);
+            return {-1, -1, nullptr, nullptr};
         }
```
[the remark this answers is not in the document]
Comment (zmnobug) [reply src/a2a3/runtime/host_build_graph/runtime/pto_ring_buffer.h:122]: Fixed in e7e6b96a. Both allocator copies now reject only when `local_task_id_ >= window_size_`, the cached-Graph preflight uses the same boundary, and the obsolete HBG scope-reclaim precheck no longer rejects the last slot first. The allocator regression proves N successes followed by an immediate N+1 failure with `current_index` unchanged; a shared a2a3/a5 Graph test also proves cached Graph submission can occupy slot N-1.

File: tests/ut/cpp/a2a3/test_hbg_tensormap.cpp [hunk 3]
```diff
@@ -70,56 +73,66 @@ class HbgTensorMapTest : public ::testing::Test {
...
 
-    // Only task 0's entry is freed; task WINDOW_SIZE's entry survives.
-    EXPECT_EQ(tmap.valid_count(), 1);
+    EXPECT_EQ(tmap.valid_count(), 2);
     TestLookupResult result;
     run_lookup(tmap, t, result);
-    ASSERT_EQ(result.count, 1);
-    EXPECT_EQ(result.entries[0].entry->producer_task_id, PTO2TaskId::make(0, WINDOW_SIZE));
+    EXPECT_EQ(result.count, 2);
+}
```
[the remark this answers is not in the document]
Comment (zmnobug) [reply tests/ut/cpp/a2a3/test_hbg_tensormap.cpp:115]: Fixed in e7e6b96a. The slot-alias test now collects the returned producer IDs and independently asserts that both task 0 and task `WINDOW_SIZE` remain present.

**Comment by zmnobug** [comment]
Addressed the review blockers in e7e6b96a:

- allow all N configured task-window slots in both architectures and cached Graph submission; remove the obsolete HBG scope-reclaim admission check;
- add immediate HBG inline-fanin diagnostics with PTO2_MAX_FANIN=128 and no PTO2_RING_DEP_POOL recommendation;
- make shared error descriptions/hints consistent with HBG task-window, heap, fanin, and TensorMap capacity failures;
- clean stale reclamation comments/constants and assert both TensorMap alias producer IDs;
- rebase the PR onto current upstream/main and keep it as one commit.

Validation:
- pre-commit: all hooks passed;
- C++ UT: 101/101 passed (the socket test was rerun outside the restricted sandbox);
- a2a3sim HBG: 26 passed, 7 skipped;
- a5sim HBG: 18 passed.

**Comment by zmnobug** [comment]
Rebased the consolidated fix onto main and force-pushed fe504d49 (supersedes e7e6b96a). The new main Graph Definition sharing path is preserved, including definition_hash/submission-only images, while the HBG final-slot preflight and forward-only capacity model remain intact.

Post-rebase validation:
- pre-commit: all hooks passed
- targeted C++ tests: 7/7 passed (allocator, TensorMap, Graph submission/cache on a2a3+a5, error names)
- a2a3sim Graph execution: 3/3 passed
- a5sim Graph execution: 3/3 passed (with the repository simulator libstdc++ preload; the first environment-only GLIBCXX mismatch was rerun successfully)

GitHub reports the PR mergeable; CI has been retriggered on fe504d49.

**Comment by zmnobug** [comment]
Follow-up CI fix pushed as 6cdef3d2 (supersedes fe504d49). profiling-flags-smoke exposed two stale host-side profiling references to the removed HBG sync_cycle and alloc_wait_cycle fields; both a2a3 and a5 runtime_maker tables now omit those retired metrics.

Local verification after the fix:
- pre-commit on both changed files: passed
- profiling orch build: a2a3sim + a5sim, HBG + T&R all passed
- profiling all-on build: a2a3sim + a5sim, HBG + T&R all passed

The prior CI run had all other jobs green, including both onboard architectures, simulator matrices, unit tests, packaging, and pre-commit.

**Comment by zmnobug** [comment]
Rebased again onto current main (7fa3f543) to resolve conflicts from the new HBG Graph execution-storage changes; pushed a0342a22 (supersedes 6cdef3d2).

Conflict resolution keeps both behaviors:
- Graph allocates owned_heap = required_heap + execution_storage_bytes from #1884/#1885.
- HBG uses all configured task-window slots (active_count >= window_size), with no obsolete check_scope_can_accept_task path.

Post-rebase validation:
- 9/9 targeted C++ tests passed, including Graph cache/activation/submission on a2a3+a5.
- all-on profiling builds passed for a2a3sim+a5sim, HBG+T&R.
- full HBG simulation passed: a2a3sim 13 passed/7 skipped; a5sim 9 passed.
- full PR-range pre-commit passed.
